### PR TITLE
Refactor so that dependency edge is added unconditionally

### PR DESCRIPTION
**Pull request by aalekseyev**
@snowleopard, I'm not entirely satisfied with this refactor, but I think it's net slightly positive.
If you like it, we can merge it. If not, feel free to take inspiration (on what not to do).

Signed-off-by: Arseniy Alekseyev <aalekseyev@janestreet.com>

### CODE REVIEW

File: src/memo/memo.ml [hunk 1]
```diff
@@ -632,8 +635,24 @@ let dep_node (type i o f) (t : (i, o, f) t) inp =
     Store.set t.cache inp dep_node;
     dep_node
 
+module Cache_lookup_result = struct
+  type ('a, 'ivar) t =
+    | New_attempt of Running_state.t * 'ivar
+    | Waiting of Running_state.t * 'ivar
```
Comment (snowleopard) [inline src/memo/memo.ml]: Looks like you're only ever accessing the `sample` field of `Running_state.t`, so you can just store `Dag.node` here directly.

Comment (snowleopard) [reply src/memo/memo.ml]: Same in the `New_attempt` constructor.

Comment (aalekseyev) [reply src/memo/memo.ml]: I thought it's distasteful to not return the information the caller was about to need, so I instead made them use it in my last commit. We now pass `running_state` to wherever it's needed.

Comment (aalekseyev) [reply src/memo/memo.ml]: In the `Waiting` case it still was unused so I made the change to store `Dag.node` there.

Comment (snowleopard) [reply src/memo/memo.ml]: Looks good!

**Pull request merged**